### PR TITLE
Reduce allocation in Index/Range.ToString

### PR DESCRIPTION
**Pull request by stephentoub**
Index.ToString first calls ToString on the integer value, and then if it's FromEnd, concatenates with "^", resulting in an extra string allocation in the FromEnd case. Range.ToString then builds on that, not only generating the intermediate Index strings, but boxing those Index structs to call string.Concat(object, object, object).

This change removes the extra allocation, so there's at most one allocation for Index/Range.ToString, the resulting string itself.  For Index, there might still be zero allocations in the case where it just delegates to Value.ToString(), and it's able to return a cached string for a single-digit integer value.

```C#
private Index _indexCached = new Index(1, false);
private Index _indexFromStart = new Index(42, false);
private Index _indexFromEnd = new Index(42, true);
private Range _rangeFromStart = Range.Create(new Index(42, false), new Index(43, false));
private Range _rangeFromEnd = Range.Create(new Index(43, true), new Index(42, true));

[Benchmark] public string IndexCached() => _indexCached.ToString();
[Benchmark] public string IndexFromStartToString() => _indexFromStart.ToString();
[Benchmark] public string IndexFromEndToString() => _indexFromEnd.ToString();
[Benchmark] public string RangeFromStartToString() => _rangeFromStart.ToString();
[Benchmark] public string RangeFromEndToString() => _rangeFromEnd.ToString();
```

Before:
```
                 Method |      Mean |     Error |    StdDev |  Gen 0 | Allocated |
----------------------- |----------:|----------:|----------:|-------:|----------:|
            IndexCached |  14.41 ns | 0.3629 ns | 0.3883 ns |      - |       0 B |
 IndexFromStartToString |  21.62 ns | 0.5399 ns | 1.5750 ns | 0.0076 |      32 B |
   IndexFromEndToString |  35.04 ns | 0.7046 ns | 0.6591 ns | 0.0153 |      64 B |
 RangeFromStartToString |  73.13 ns | 1.7092 ns | 2.2224 ns | 0.0362 |     152 B |
   RangeFromEndToString | 108.04 ns | 2.0858 ns | 1.8490 ns | 0.0516 |     216 B |
```

After:
```
                 Method |     Mean |     Error |    StdDev |  Gen 0 | Allocated |
----------------------- |---------:|----------:|----------:|-------:|----------:|
            IndexCached | 14.19 ns | 0.0979 ns | 0.0818 ns |      - |       0 B |
 IndexFromStartToString | 22.70 ns | 0.1632 ns | 0.1363 ns | 0.0076 |      32 B |
   IndexFromEndToString | 34.87 ns | 0.6725 ns | 0.5962 ns | 0.0076 |      32 B |
 RangeFromStartToString | 55.33 ns | 0.3032 ns | 0.2192 ns | 0.0095 |      40 B |
   RangeFromEndToString | 55.92 ns | 0.3618 ns | 0.3021 ns | 0.0095 |      40 B |
```

cc: @tarekgh, @jkotas

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/Index.cs [hunk 1]
```diff
@@ -28,10 +28,14 @@ public override int GetHashCode()
...
+
+        private string ToStringFromEnd()
         {
-            string str = Value.ToString();
-            return FromEnd ? "^" + str : str;
+            Span<char> span = stackalloc char[11]; // 1 for ^ and 10 for longest possible uint value
+            ((uint)Value).TryFormat(span.Slice(1), out int charsWritten);
+            span[0] = '^';
+            return new string(span.Slice(0, charsWritten + 1));
         }
```
Comment (EgorBo) [inline src/System.Private.CoreLib/shared/System/Index.cs]: As an alternative - modified version of `UInt32ToDecStr` without CountDigits, etc
```csharp
private static unsafe string UInt32ToDecStr(uint value)
{
    char* buf = stackalloc char[11];
    int i = 11;
    do
    {
        uint div = value / 10;
        buf[--i] = (char)('0' + value - (div * 10));
        value = div;
    }
    while (value != 0);
    buf[--i] = '^';
    return new string(buf, i, 11 - i);
}
```

Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Index.cs]: @stephentoub yeah, surely doesn't make any sense, I was actually curious if your technique with `stackalloc[11]` could be used in `UInt32ToDecStr`, but looks like GetDigitsCount+string.FastAllocate is better 🙂

**Comment by stephentoub** [comment]
> Nit: should we have a debug assert after the calls to TryFormat? I could go either way on this.

Thanks, Levi.  I considered that, too, but was on the fence and ended up removing them to avoid the clutter.  Since you also suggested it, I'll add them back.

**Comment by stephentoub** [comment]
@dotnet-bot test Windows_NT arm Cross Debug Innerloop Build please
@dotnet-bot test Windows_NT arm64 Cross Debug Innerloop Build please

**Pull request merged**